### ISSUE TITLE
serialization breaks in python 2.7.3

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#14893]
## 🐛 Bug

Running unit tests, test_pickle breaks:

```
======================================================================
ERROR: test_pickle (__main__.TestAutograd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_autograd.py", line 1914, in test_pickle
    serialized = [pickle.dumps([x, y], protocol=p) for p in range(3)]
  File "/Users/soumith/miniconda2/lib/python2.7/site-packages/torch/storage.py", line 34, in __reduce__
    torch.save(self, b)
  File "/Users/soumith/miniconda2/lib/python2.7/site-packages/torch/serialization.py", line 218, in save
    return _with_file_like(f, "wb", lambda f: _save(obj, f, pickle_module, pickle_protocol))
  File "/Users/soumith/miniconda2/lib/python2.7/site-packages/torch/serialization.py", line 143, in _with_file_like
    return body(f)
  File "/Users/soumith/miniconda2/lib/python2.7/site-packages/torch/serialization.py", line 218, in <lambda>
    return _with_file_like(f, "wb", lambda f: _save(obj, f, pickle_module, pickle_protocol))
  File "/Users/soumith/miniconda2/lib/python2.7/site-packages/torch/serialization.py", line 297, in _save
    serialized_storages[key]._write_file(f, _should_read_directly(f))
TypeError: 'buffer' does not have the buffer interface

----------------------------------------------------------------------
```

## To Reproduce

1. Install on 2.7.3 - OSX
1. run `python run_test.py`
1.

cc @ezyang @gchanan

**Comment by vince62s** [comment]
Any update on this ?

**Comment by zou3519** [comment]
@vince62s this might be a python 2.7.3 bug and the particular python version in question is now unsupported. Can you upgrade your Python?

**Comment by vince62s** [comment]
I don't think so, we got it with 2.7.13

**Comment by vince62s** [comment]
if you have to time read this thread, it will show you the error and how we ended up to find a workaround:
https://github.com/OpenNMT/OpenNMT-py/pull/1188

**Comment by zou3519** [comment]
Okay, let me try to reproduce this first. @vince62s did you get this from running test_pickle on python 2.7.13 or were you running something else?

**Comment by flauted** [comment]
I'm the author of the PR @vince62s linked to above. I don't think that this issue is related.

To summarize that thread, we were ``torch.save``ing a subclass of ``torchtext.data.Dataset``. The subclass defined a ``__reduce_ex__(self, proto)`` method that forwarded to ``super(OurDataset, self).__reduce_ex__()`` without any arguments. That worked on 2.7.13 and Python 3.6.8 - surprisingly, since the pickle docs say ``__reduce_ex__`` takes an argument - but broke on Python 3.7.2.

The root cause seemed to be: ``torchtext.data.Dataset`` defines a custom ``__getattr__`` as a **generator** that yields the attr from all its examples, and in Python 2.7 the ``pickle`` module (and, presumably, the cPickle module) looks for ``__reduce_ex__`` with ``getattr`` and then calls it. Well, generators aren't callable.

----
To illustrate, this is kind of what was going on:
```
class Base(object):
    def __init__(self):
        self.alpha = ["a", "b", "c", "d"]

    def __getattr__(self, attr):
        if attr in self.alpha:
            for x in self.alpha:
                yield getattr(x, attr)


class Sub(Base):
    pass


obj = Sub()
getattr(obj, "__reduce_ex__")(2)  # this is basically what pickle was doing
```
```
TypeError: 'generator' object is not callable
```
And the fix was:
```
class Sub(Base):
    def __getattr__(self, attr):
        # saveguard against infinite recursion when 'alpha' DNE
        if 'alpha' not in vars(self):
            raise AttributeError
        if attr in self.alpha:
            # RETURN a generator
            return (getattr(x, attr) for x in self.alpha)
        else:
            raise AttributeError

obj = Sub()
getattr(obj, "__reduce_ex__")(2)
```
And with that we make it into ``object.__reduce_ex__``, which is what we wanted.
```
(<function copy_reg.__newobj__>,
 (__main__.Sub,),
 {'alpha': ['a', 'b', 'c', 'd']},
 None,
 None)
```

**Comment by skrah** [comment]
With Python 2.7.16 compiled from source and `master` I cannot reproduce this.  I'm on Linux though:

```
$ /home/stefan/2.7/bin/python --version
Python 2.7.16
$ /home/stefan/2.7/bin/python -m pytest -v -x -k pickle test_autograd.py
[...]
test_autograd.py::TestAutograd::test_pickle PASSED
```

**Comment by skrah** [comment]
@vince62s (or anyone else) Can you still reproduce the specific `test_pickle` error under the conditions I mentioned above (Python 2.7.16 compiled from source, no conda)?

**Comment by rgommers** [comment]
I have tried to reproduce this on macOS 10.14 (Mojave, most recent version), with all combinations of:

- Python 2.7.13 as well as 2.7.16 from the conda defaults channel
- PyTorch 1.2.0 and latest nightly

This doesn't reproduce anymore. 

I also looked at https://github.com/OpenNMT/OpenNMT-py/pull/1188 and agree with @flauted that that's unrelated.

No other reports in the last 8 months (except for the unrelated OpenNMT one), this seems fixed. I propose to close the issue.
